mathml-block.php

```diff
@@ -47,29 +84,98 @@ function mathml_set_up_js_translations() {
 add_action( 'init', __NAMESPACE__ . '\mathml_set_up_js_translations' );
 
 /**
- * Potentially enqueue the front end mathjax script, if any mathml blocks are detected in the content.
+ * Register block.
  */
-function potentially_add_front_end_mathjax_script() {
-	global $post;
-
-	// Only apply on singular pages.
-	if ( ! is_singular() ) {
+function register_block() {
+	if ( ! function_exists( 'register_block_type' ) ) {
 		return;
 	}
 
-	// Check the content for mathml blocks.
-	$has_mathml_block  = strpos( $post->post_content, 'wp:mathml/mathmlblock' );
-	$has_mathml_inline = strpos( $post->post_content, '<mathml>' );
-	if ( false === $has_mathml_block && false === $has_mathml_inline ) {
-		return;
+	$registry = WP_Block_Type_Registry::get_instance();
+
+	// @todo This can probably be de-duplicated in the JS code with registerBlockType.
+	$attributes = array(
+		'formula' => array(
+			'source'   => 'html',
+			'selector' => 'div',
+			'type'     => 'string',
+		),
+	);
+
+	if ( $registry->is_registered( BLOCK_NAME ) ) {
+		$block                  = $registry->get_registered( BLOCK_NAME );
+		$block->render_callback = __NAMESPACE__ . '\render_block';
+		$block->attributes      = array_merge( $block->attributes, $attributes );
+	} else {
+		register_block_type(
+			BLOCK_NAME,
+			[
+				'render_callback' => __NAMESPACE__ . '\render_block',
+				'attributes'      => $attributes,
+			]
+		);
 	}
+}
+add_action( 'init', __NAMESPACE__ . '\register_block' );
 
-	// Filter the MathJax config string.
-	$config_string = apply_filters( 'mathml_block_mathjax_config', 'TeX-MML-AM_CHTML' );
+/**
+ * Add async attribute to MathJax script tag.
+ *
+ * @param string $tag    Script tag.
+ * @param string $handle Script handle.
+ *
+ * @return string Script tag.
+ */
+function add_async_to_mathjax_script_loader_tag( $tag, $handle ) {
+	if ( MATHJAX_SCRIPT_HANDLE === $handle ) {
+		$tag = preg_replace( '/(?<=<script\s)/', ' async ', $tag );
+	}
+	return $tag;
+}
+
+/**
+ * Render block.
+ *
+ * Creates an <amp-mathml> element on AMP responses.
+ *
+ * @param array  $attributes Attributes.
+ * @param string $content    Content.
+ *
+ * @return string Rendered block.
+ */
+function render_block( $attributes, $content = '' ) {
+	if ( is_admin() || ! preg_match( '#^(?P<start_div>\s*<div.*?>)(?P<formula>.+)(?P<end_div></div>\s*)$#s', $content, $matches ) ) {
```

Comment:
What does this regex do? The goal is to only enqueue the mathjax script on the front end when there are <mathml> (inline) tags or mathml blocks.

Reply:
The regex is for parsing the `$content`, in particular to isolate the formula from the start tag and end tag for the enclosed `div`. This is done not only to extract the formula to supply into the `amp-mathml` tag, but also in the non-AMP version to be able to inject the scripts _inside_ the block wrapper. This ensures that styles that target blocks that have a certain sibling won't get tripped up by the injected `script`. The scripts are printed right away instead of waiting for `wp_footer` in order to minimize the flash of unstyled content.

Reply:
Apparently the regex needs to be adapted to support inline `<mathml>` tags per below.